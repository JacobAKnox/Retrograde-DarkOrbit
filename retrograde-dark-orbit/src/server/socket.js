"use client"

import { navigate } from './../components/navigation';
import { useEffect } from 'react';
// import this module where you need to handle
// socket.io interactions :)
import { io } from 'socket.io-client';

const server_addr = process.env.NEXT_PUBLIC_SERVERADDRESS || "localhost";
const server_port = process.env.NEXT_PUBLIC_SERVERPORT || "4000";

console.log(server_addr);

let socket = io(`http://${server_addr}:${server_port}`);
let recMessage = (e) => {};

const connect = () => {
    if (typeof window !== 'undefined') {
        // session storage to make testing easier, probably change to local storage later
        // session storage is not shared across tabs, but is across refresh
        const sessionID = sessionStorage.getItem("sessionID");

        if (sessionID) {
            socket.auth = { sessionID };
        }
    }

    socket.on("session", ({ sessionID, userID }) => {
        socket.auth = { sessionID };
        sessionStorage.setItem("sessionID", sessionID);
        socket.userID = userID;
    });

    socket.on("game_start", ({code}) => {
        navigate(`/game?code=${code}`);
    });

    socket.connect();
}

export async function join_lobby(username, code) {
    return await socket.emitWithAck('join', {username: username, code: code});
}

export async function leave_lobby() {
    return await socket.emitWithAck('leave');
}

socket.on("receive chat msg", (message) => {
    console.log(message, socket.id);
})

export async function create_lobby(username) {
    return await socket.emitWithAck('create', {username});
}

export default function Connector() {
    useEffect(connect, []);
}

export async function chat_message(message) {
    socket.emit("send chat msg", {message});
}

export function chat_message_listener(callback) {
    recMessage = callback;
}

export function update_player_ready() {
    socket.emit("player_ready");
}

export const update_ready_status = (updateReadyStatus) => {
  socket.on("ready_count_updated", (data) => {
    updateReadyStatus({
      num_ready: data.readyCount, 
      num_total: data.totalPlayers
    });
  });
  socket.emit("init ready count");
};

socket.on("receive chat msg", ({username, message}) => {
    recMessage('[' + username + ']: ' + message)
})

// socket.on("lobby code", (code) => {
//     console.log("FROM SOCKET ON");
//     console.log(code);
// })
