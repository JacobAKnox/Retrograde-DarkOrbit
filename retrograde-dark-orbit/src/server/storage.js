// uses session storage which is not presevreed across tabs, but is across refresh
// switch to localstorage at some point
export function getItem(itemId) {
    if (typeof window !== 'undefined') {
        return sessionStorage.getItem(itemId);
    }
    return undefined;
}

export function storeItem(itemId, item) {
    if (typeof window !== 'undefined') {
        sessionStorage.setItem(itemId, item);
    }
}