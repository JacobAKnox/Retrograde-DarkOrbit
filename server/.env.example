MONGODB_URI="mongodb://localhost:27017"
DATABASE="RetrogradeDarkOrbit"
PORT=4000