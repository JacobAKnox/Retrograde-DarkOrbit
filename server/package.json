{
  "name": "socket-chat-example",
  "version": "0.0.1",
  "description": "my first socket.io app",
  "type": "module",
  "scripts": {
    "test": "jest"
  },
  "dependencies": {
    "express": "^4.18.2",
    "socket.io": "^4.7.4",
    "socket.io-client": "^4.7.4"
  },
  "devDependencies": {
    "@babel/preset-env": "^7.23.9",
    "jest": "^29.7.0"
  }
}
