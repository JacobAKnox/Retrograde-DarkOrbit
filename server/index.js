import express from "express";
import {createServer} from "node:http";
import {Server} from "socket.io";
import { join_lobby, create_lobby, leave_lobby, get_lobby, get_num_ready_players, get_num_players } from "./lobbies/lobbies.js";
import { find_or_create_session } from "./sessions/sessions.js";
import { assign_roles, get_game, get_role_info, setup, start_game, validate_received_user_poi_values, get_player_POIs, set_player_POIs } from "./games/game.js";
import { set_player_ready } from "./lobbies/lobbies.js";
import { PHASE_STATES } from "./games/game_globals.js";
import { gameLoop, set_status_bar_update, set_timer_update_callback, set_ids_and_names_callback, winners_update } from "./games/turns.js";

const app = express();
const server = createServer(app);
export const io = new Server(server, {
  cors: {
    origin: '*',
  }
});

function redirect_user(socket) {
  if (socket.roomCode === "") {
    socket.emit("redirect", "/");
    return;
  }
  if (get_game(socket.roomCode)) {
    socket.emit("redirect", `/game?code=${socket.roomCode}`);
    return;
  }
  socket.emit("redirect", `/lobby?code=${socket.roomCode}`);
}

io.use((socket, next) => {
  const sessionID = socket.handshake.auth.sessionID;
  const result = find_or_create_session(sessionID);
  socket.sessionID = result.sessionId;
  socket.userID = result.userId;
  socket.roomCode = result.code;
  socket.username = result.username;
  if (socket.roomCode !== "") {
    socket.join(socket.roomCode);
  }
  setTimeout(() => redirect_user(socket), 500);
  next();
});

io.on("connection", (socket) => {
  console.log("user connected");

  // text chat
  socket.on("send chat msg", ({message}) => {
    console.log('[Room:' + socket.roomCode + ' chat] ' + socket.username + ': ' + message);
    io.in(socket.roomCode).emit("receive chat msg", {username: socket.username, message});
  });

  

  // POI updates during action phase
  socket.on("client-sent poi update", (POIs, callback) => {
    console.log('[Room: ' + socket.roomCode + ', User: ' + socket.username + ', POI update]:');
    const allowed_phases = [PHASE_STATES.DISCUSSION_PHASE, PHASE_STATES.ACTION_PHASE];
    let game = get_game(socket.roomCode);
    if (Object.keys(POIs).length === 0) { 
      console.log("AAAAAAAAAAAAAAAAAAAAAAAAAA");
    }
    if (!game) {
      callback({
        status: 404,
        message: "You are not in a game"
      });
      return;
    }

    if (!allowed_phases.includes(game.currentState)) {
      callback({
        status: 405,
        message: "cannot update point allocation during this phase"
      });
      if (Object.keys(get_player_POIs(game, socket.userID)).length === 0) { 
        console.log("AAAAAAAAAAAAAAAAAAAAAAAAAA");
      }
      socket.emit("server-sent poi update", get_player_POIs(game, socket.userID));
      return;
    }

    if(!validate_received_user_poi_values(game, socket.userID, POIs)) {
      callback({
        status: 409,
        message: "client POIs not valid"
      });
      console.log("POIs BAG FAILED!");
      socket.emit("server-sent poi update", get_player_POIs(game, socket.userID));
    }
    else {
      callback({
        status: 200,
        message: "POIs OK"
      });
      set_player_POIs(game, socket.userID, POIs);
    }
  });

  socket.on("join", (data, callback) => {
    if (data.code === undefined || data.username === undefined) {
      // this shouldn't happen unless someone is doing something outside the website
      callback({
        status: 400,
        message: "bad packet"
      });
    }
    const result = join_lobby(data.code, data.username, socket.userID);
    if (result.status === 200) {
      socket.join(data.code);
      socket.username = data.username;
      socket.roomCode = data.code;
      updatePlayerList(socket.roomCode); //added this
    }
    callback(result);
  });

  socket.on("leave", (callback) => {
    console.log("left");
    callback(leave_lobby(socket.userID));
    updatePlayerList(socket.roomCode); //added this
  });

  socket.emit("session", {
    sessionID: socket.sessionID,
    userID: socket.userID
  });

  // socket.emit("lobby code", socket.roomCode);

  socket.on("disconnect", () => {
    console.log("user disconnected")
    updatePlayerList(socket.roomCode); //added this 
  });

  socket.on("create", (data, callback) => {
    if (data.username === undefined || socket.userID === null) {
      // this shouldn't happen unless someone is doing something outside the website
      callback({
        status: 400,
        message: "bad packet"
      });
    }

    const result = create_lobby(data.username, socket.userID);
    if (result.status === 200) {
      socket.join(result.code);
      socket.username = data.username;
      socket.roomCode = result.code;
    }
    callback(result);
  });
  
  // socket.emit("lobby code", "socket.roomCode");

  socket.on("player_ready",() => {
    const userID = socket.userID;
    const result = set_player_ready(userID);
    if (result.status === 200) {
      const lobby = get_lobby(socket.roomCode);
      updatePlayerList(socket.roomCode); //added this
      io.in(socket.roomCode).emit("ready_count_updated", { 
        readyCount: get_num_ready_players(socket.roomCode), 
        totalPlayers: Object.keys(lobby).length 
      });
    }
    try_start_game(socket);
  });



  async function try_start_game(socket) {
    if (socket.roomCode === "") {
      return;
    }
    const lobby = get_lobby(socket.roomCode);
    if (!lobby) {
      return;
    }
    if (get_num_ready_players(socket.roomCode) < get_num_players(socket.roomCode)) {
      // not enough players ready
      return;
    }

    const result = start_game(lobby, socket.roomCode);
    if (result.status !== 200) {
      // notify clients that the game start has failed.
      io.in(socket.roomCode).emit("receive chat msg", {username: "server", message: `failed to start game. \n ${result.message}`});
    }

    let game = get_game(socket.roomCode);
    assign_roles(game);

    // tell clients to start the game
    io.in(socket.roomCode).emit("game_start", {code: socket.roomCode});
    
    // tell each player their role
    // delay telling players thier role so that they have time to load the page
    setTimeout(async () => {
      const sockets = await io.in(socket.roomCode).fetchSockets();
      sockets.forEach(s => {
        s.emit("role_info", get_role_info(game, s.userID));
      });
    }, 1000);
    gameLoop(socket.roomCode);
  }

  socket.on("init ready count", () => {
      const lobby = get_lobby(socket.roomCode);
      io.in(socket.roomCode).emit("ready_count_updated", { 
      readyCount: get_num_ready_players(socket.roomCode), 
      totalPlayers: Object.keys(lobby).length 
    });
  });

  socket.on('request_player_list', () => {
    updatePlayerList(socket.roomCode);
  });

});

const PORT = process.env.PORT | 4000;
server.listen(PORT, async () => {
  await setup();
  set_timer_update_callback(updateTimer);
  set_ids_and_names_callback(sendIdsAndNames);
  set_status_bar_update(updateStatusBar);
  winners_update(sendWinnersToClient);
  console.log(`server running at http://localhost:${PORT}`);
});

export function closeServer() {
  server.close();
}

export function updateTimer(phase, time, lobbyCode){
  io.in(lobbyCode).emit("update timer phase", {length: time, name: phase});
} 

export function sendIdsAndNames(IDSANDNAMES, lobbyCode){
  io.in(lobbyCode).emit("server-sent poi update", IDSANDNAMES);
} 

function updateStatusBar(lobbyCode, statusBars) {
  io.in(lobbyCode).emit("status_update", statusBars);
}

function sendWinnersToClient(lobbyCode, winners) {
  io.in(lobbyCode).emit("winner_data", winners);
}
//update player lplayer
function updatePlayerList(lobbyCode) {
  const lobby = get_lobby(lobbyCode);
  console.log(`Lobby retrieved: ${JSON.stringify(lobby)}`);
  
  if (!lobby || !Object.keys(lobby).length) {
      console.error(`No players in lobby: ${lobbyCode} or lobby does not exist.`);
      return;
  }

  const playerList = Object.values(lobby).map(player => ({
      id: player.id, 
      name: player.username,
      ready: player.ready_state 
  }));

  console.log(`Emitting player list for lobby ${lobbyCode}:`, playerList); 
  io.in(lobbyCode).emit('player_list_updated', playerList);
}
