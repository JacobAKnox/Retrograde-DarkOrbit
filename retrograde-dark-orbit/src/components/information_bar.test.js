
jest.mock('../server/socket', () => ({
  listen_status_bar_update: jest.fn().mockImplementation((callback) => {
    callback({
      "crew": {name: "Crew", value: 80, max_value: 100},
      "ship_health": {name: "Ship Health", value: 90, max_value: 100},
      "fuel": {name: "Fuel", value: 70, max_value: 100},
      "life_support": {name: "Life Support", value: 60, max_value: 100},
      "power": {name: "Power", value: 50, max_value: 100}
    });
  })
}));

import React from 'react';
import { render, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom'; 
import InformationBar from './information_bar';
import {listen_status_bar_update} from '../server/socket';

describe('<InformationBar />', () => {

  beforeEach(() => {
    listen_status_bar_update.mockClear();
  });

  it('renders correctly', () => {

    const { getByText } = render(<InformationBar />);

    expect(getByText(/crew/i)).toBeInTheDocument();
    expect(getByText(/ship health/i)).toBeInTheDocument();
    expect(getByText(/fuel/i)).toBeInTheDocument();
    expect(getByText(/life support/i)).toBeInTheDocument();
    expect(getByText(/power/i)).toBeInTheDocument();

  });


  it('updates progress bars when socket emits new game status', async () => {

    render(<InformationBar />);

    await waitFor(() => {

      const progressBarCrew = document.querySelector('progress[aria-label="Crew"]');
      expect(progressBarCrew).toHaveAttribute('value', '80');

      const progressBarHealth = document.querySelector('progress[aria-label="Ship Health"]');
      expect(progressBarHealth).toHaveAttribute('value', '90');

      const progressBarFuel = document.querySelector('progress[aria-label="Fuel"]');
      expect(progressBarFuel).toHaveAttribute('value', '70');

      const progressBarLifeSupport = document.querySelector('progress[aria-label="Life Support"]');
      expect(progressBarLifeSupport).toHaveAttribute('value', '60');

      const progressBarPower = document.querySelector('progress[aria-label="Power"]');
      expect(progressBarPower).toHaveAttribute('value', '50');

    });
  });
});
