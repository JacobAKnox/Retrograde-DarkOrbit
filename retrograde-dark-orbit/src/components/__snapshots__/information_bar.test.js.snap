// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InformationBar /> renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <style>
        
                .information-bar {
                    display: flex;
                    flex-direction: column;
                    gap: 10px;
                    background: #151B30; 
                    padding: 15px;
                    border-radius: 8px;
                    color: #fff; 
                    margin: 10px 0;
                }

                .information-bar label {
                    font-size: 0.9rem;
                    color: #FFFFFF;
                }

                .information-bar progress {
                    width: 100%;
                    height: 20px;
                    border-radius: 10px;
                    color: #4caf50;
                    background-color: #4ade80;
                }
               
                .information-bar progress::-webkit-progress-value {
                    background-color: #4caf50;
                    border-radius: 10px;
                }
     
                .information-bar progress::-webkit-progress-bar {
                    background-color: #333;
                    border-radius: 10px;
                }
                
      </style>
      <div
        class="information-bar"
      >
        <label>
          Crew: 
          <div
            class="bg-slate-200 relative rounded-xl text-black text-center border-slate-600 border-solid border-2 h-6 content-center overflow-clip"
          >
            <div
              class="bg-green-700 h-5 absolute bottom-0"
              style="left: 5%; width: 95%;"
            />
            <div
              class="bg-slate-700 h-5 w-1 absolute bottom-0"
              style="left: 50%;"
            />
          </div>
           
        </label>
        <label>
          Ship Health: 
          <div
            class="bg-slate-200 relative rounded-xl text-black text-center border-slate-600 border-solid border-2 h-6 content-center overflow-clip"
          >
            <div
              class="bg-green-700 h-5 absolute bottom-0"
              style="left: 20%; width: 80%;"
            />
            <div
              class="bg-slate-700 h-5 w-1 absolute bottom-0"
              style="left: 50%;"
            />
          </div>
        </label>
        <label>
          Fuel: 
          <div
            class="bg-slate-200 relative rounded-xl text-black text-center border-slate-600 border-solid border-2 h-6 content-center overflow-clip"
          >
            <div
              class="bg-green-700 h-5 absolute bottom-0"
              style="left: 80%; width: 20%;"
            />
            <div
              class="bg-slate-700 h-5 w-1 absolute bottom-0"
              style="left: 50%;"
            />
          </div>
        </label>
        <label>
          Life Support: 
          <div
            class="bg-slate-200 relative rounded-xl text-black text-center border-slate-600 border-solid border-2 h-6 content-center overflow-clip"
          >
            <div
              class="bg-green-700 h-5 absolute bottom-0"
              style="left: 50%; width: 50%;"
            />
            <div
              class="bg-slate-700 h-5 w-1 absolute bottom-0"
              style="left: 50%;"
            />
          </div>
        </label>
        <label>
          Power: 
          <div
            class="bg-slate-200 relative rounded-xl text-black text-center border-slate-600 border-solid border-2 h-6 content-center overflow-clip"
          >
            <div
              class="bg-green-700 h-5 absolute bottom-0"
              style="left: 30%; width: 70%;"
            />
            <div
              class="bg-slate-700 h-5 w-1 absolute bottom-0"
              style="left: 50%;"
            />
          </div>
        </label>
      </div>
    </div>
  </body>,
  "container": <div>
    <style>
      
                .information-bar {
                    display: flex;
                    flex-direction: column;
                    gap: 10px;
                    background: #151B30; 
                    padding: 15px;
                    border-radius: 8px;
                    color: #fff; 
                    margin: 10px 0;
                }

                .information-bar label {
                    font-size: 0.9rem;
                    color: #FFFFFF;
                }

                .information-bar progress {
                    width: 100%;
                    height: 20px;
                    border-radius: 10px;
                    color: #4caf50;
                    background-color: #4ade80;
                }
               
                .information-bar progress::-webkit-progress-value {
                    background-color: #4caf50;
                    border-radius: 10px;
                }
     
                .information-bar progress::-webkit-progress-bar {
                    background-color: #333;
                    border-radius: 10px;
                }
                
    </style>
    <div
      class="information-bar"
    >
      <label>
        Crew: 
        <div
          class="bg-slate-200 relative rounded-xl text-black text-center border-slate-600 border-solid border-2 h-6 content-center overflow-clip"
        >
          <div
            class="bg-green-700 h-5 absolute bottom-0"
            style="left: 5%; width: 95%;"
          />
          <div
            class="bg-slate-700 h-5 w-1 absolute bottom-0"
            style="left: 50%;"
          />
        </div>
         
      </label>
      <label>
        Ship Health: 
        <div
          class="bg-slate-200 relative rounded-xl text-black text-center border-slate-600 border-solid border-2 h-6 content-center overflow-clip"
        >
          <div
            class="bg-green-700 h-5 absolute bottom-0"
            style="left: 20%; width: 80%;"
          />
          <div
            class="bg-slate-700 h-5 w-1 absolute bottom-0"
            style="left: 50%;"
          />
        </div>
      </label>
      <label>
        Fuel: 
        <div
          class="bg-slate-200 relative rounded-xl text-black text-center border-slate-600 border-solid border-2 h-6 content-center overflow-clip"
        >
          <div
            class="bg-green-700 h-5 absolute bottom-0"
            style="left: 80%; width: 20%;"
          />
          <div
            class="bg-slate-700 h-5 w-1 absolute bottom-0"
            style="left: 50%;"
          />
        </div>
      </label>
      <label>
        Life Support: 
        <div
          class="bg-slate-200 relative rounded-xl text-black text-center border-slate-600 border-solid border-2 h-6 content-center overflow-clip"
        >
          <div
            class="bg-green-700 h-5 absolute bottom-0"
            style="left: 50%; width: 50%;"
          />
          <div
            class="bg-slate-700 h-5 w-1 absolute bottom-0"
            style="left: 50%;"
          />
        </div>
      </label>
      <label>
        Power: 
        <div
          class="bg-slate-200 relative rounded-xl text-black text-center border-slate-600 border-solid border-2 h-6 content-center overflow-clip"
        >
          <div
            class="bg-green-700 h-5 absolute bottom-0"
            style="left: 30%; width: 70%;"
          />
          <div
            class="bg-slate-700 h-5 w-1 absolute bottom-0"
            style="left: 50%;"
          />
        </div>
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
